File: leetcode/BestTimetoBuyandSellStock/maxProfit.java
public class Solution {
    public int maxProfit(int[] prices) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int profit = 0, min = Integer.MAX_VALUE;
        for (int x : prices) {
            if (x < min) min = x;
            profit = Math.max(x - min, profit);
        }
        return profit;
    }
}

File: leetcode/BestTimetoBuyandSellStockWithCooldown/maxProfit.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int size = prices.size();
        if (!size) return 0;
        vector<int> sol(size, 0);
        for (int i = size - 2; i >= 0; i--) {
            int minPrice = 0;
            for (int j = i; j < size; j++) {
                if (j == i) minPrice = prices[j];
                else {
                    if (prices[j] > minPrice) {
                        sol[i] = max(sol[i], prices[j] - minPrice + (j + 2 < size ? sol[j + 2] : 0));
                    }
                    minPrice = min(minPrice, prices[j]);
                }
            }
        }
        return sol[0];
    }
};
